lib_eio_posix/eio_posix_stubs.c

```diff
@@ -0,0 +1,178 @@
+#define _FILE_OFFSET_BITS 64
+
+#include <sys/types.h>
+#include <sys/random.h>
+#include <errno.h>
+#include <sys/uio.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+
+#ifndef __linux__
+# include <stdlib.h>
+#endif
+
+#include <caml/mlvalues.h>
+#include <caml/memory.h>
+#include <caml/alloc.h>
+#include <caml/unixsupport.h>
+#include <caml/bigarray.h>
+
+#ifdef ARCH_SIXTYFOUR
+#define Int63_val(v) Long_val(v)
+#else
+#define Int63_val(v) (Int64_val(v)) >> 1
+#endif
+
+CAMLprim value caml_eio_posix_getrandom(value v_ba, value v_off, value v_len) {
+  CAMLparam1(v_ba);
+  ssize_t ret;
+  ssize_t off = (ssize_t)Long_val(v_off);
+  ssize_t len = (ssize_t)Long_val(v_len);
+  do {
+    void *buf = Caml_ba_data_val(v_ba) + off;
+    caml_enter_blocking_section();
+#ifdef __linux__
+    ret = getrandom(buf, len, 0);
+#else
+    arc4random_buf(buf, len);
+    ret = len;
+#endif
+    caml_leave_blocking_section();
+  } while (ret == -1 && errno == EINTR);
+  if (ret == -1) uerror("getrandom", Nothing);
+  CAMLreturn(Val_long(ret));
+}
+
+/* Fill [iov] with pointers to the cstructs in the array [v_bufs]. */
+static void fill_iov(struct iovec *iov, value v_bufs) {
+  int n_bufs = Wosize_val(v_bufs);
+  for (int i = 0; i < n_bufs; i++) {
+    value v_cs = Field(v_bufs, i);
+    value v_ba = Field(v_cs, 0);
+    value v_off = Field(v_cs, 1);
+    value v_len = Field(v_cs, 2);
+    iov[i].iov_base = Caml_ba_data_val(v_ba) + Long_val(v_off);
+    iov[i].iov_len = Long_val(v_len);
+  }
+}
+
+CAMLprim value caml_eio_posix_readv(value v_fd, value v_bufs) {
+  CAMLparam1(v_bufs);
+  ssize_t r;
+  int n_bufs = Wosize_val(v_bufs);
+  struct iovec iov[n_bufs];
+
+  fill_iov(iov, v_bufs);
+
+  r = readv(Int_val(v_fd), iov, n_bufs);
+  if (r < 0) uerror("readv", Nothing);
+
+  CAMLreturn(Val_long(r));
+}
+
+CAMLprim value caml_eio_posix_writev(value v_fd, value v_bufs) {
+  CAMLparam1(v_bufs);
+  ssize_t r;
+  int n_bufs = Wosize_val(v_bufs);
+  struct iovec iov[n_bufs];
+
+  fill_iov(iov, v_bufs);
+
+  r = writev(Int_val(v_fd), iov, n_bufs);
+  if (r < 0) uerror("writev", Nothing);
+
+  CAMLreturn(Val_long(r));
+}
+
+CAMLprim value caml_eio_posix_preadv(value v_fd, value v_bufs, value v_offset) {
+  CAMLparam2(v_bufs, v_offset);
+  ssize_t r;
+  int n_bufs = Wosize_val(v_bufs);
+  struct iovec iov[n_bufs];
+
+  fill_iov(iov, v_bufs);
+
+  r = preadv(Int_val(v_fd), iov, n_bufs, Int63_val(v_offset));
+  if (r < 0) uerror("preadv", Nothing);
+
+  CAMLreturn(Val_long(r));
+}
+
+CAMLprim value caml_eio_posix_pwritev(value v_fd, value v_bufs, value v_offset) {
+  CAMLparam2(v_bufs, v_offset);
+  ssize_t r;
```

Comment:
Hmm… Is there a reason to prefer ancient C style with declarations at the beginning of a block?

I would recommend against that unless there is some really good reason for it.

Reply:
Historically C99 declarations didn't get that much adoption in the open source world (especially the UNIX world) because portability was important, I don't think this holds in 2023 and I'd be surprised if there is any issue. Having said that, I dislike declarations in the middle of a block, I like the discipline of having to declare beforehand.

Reply:
Yeah, let's keep style preferences out of the review (I do prefer the newer style, but most other OCaml code seems to be written this way).

Reply:
The reason C code used to be written that way was that it was required.

Nowadays C supports having declarations anywhere that a statement can be used.  This allows one to make it so that all variables are initialized with a proper value.  It reduces the chance of making programming errors such as using uninitialized variables or having type mismatches (variable declared with one type then later initialized with another).  It also promotes the use of more variables (rather than e.g. reusing a variable for multiple different purposes) and even using variables as in FP such that they are only initialized once and never written, which can further help to avoid programming errors.  It also makes it easier to read the code as the scopes of variables can be narrower and be seen more clearly.

Reply:
If we wanna catch more errors, which is always good, I'd suggest adding proper warnings to the compilation flags, the ones that dune select are absurdly spartan (as it's inherited from ocamlc_cflags, we basically build without any warning enabled, standard flags are these:
`ocamlc_cflags: -O2 -fno-strict-aliasing -fwrapv -pthread -fPIC`